**PHARMACEUTICAL DRUG NAME EXTRACTION USING NAMED ENTITY RECOGNITON**





USING EASYOCR FOR EXTRACTING TEXT FROM IMAGE


In [2]:
!pip install easyocr

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.9/2.9 MB 41.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.6/47.6 MB 19.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 619.2/619.2 KB 32.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 146.0/146.0 KB 10.0 MB/s eta 0:00:00
  Attempting uninstall: opencv-python-headless
    Found existing installation: opencv-python-headless 4.7.0.68
    Uninstalling opencv-python-headless-4.7.0.68:
      Successfully uninstalled opencv-python-headless-4.7.0.68


In [3]:
import easyocr #Used for extracting text from image which will then be sent to our ner model for medicine name detection.

In [4]:
from google.colab import files
uploaded = files.upload() #Upload the image containing the back side of medicine pamphlet which includes the medicine name and it's description


Saving aug0003_.jpg to aug0003_.jpg


In [6]:
import pandas as pd
import cv2
import easyocr
img = cv2.imread('/content/aug0003_.jpg')
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
noise=cv2.medianBlur(gray,3)
thresh = cv2.threshold(noise, 0, 255, cv2.THRESH_BINARY | cv2.THRESH_OTSU)[1]
reader = easyocr.Reader(['en'])
result = reader.readtext(img,paragraph='False')
text=pd.DataFrame(result)
text

0  \
0  [[891, 218], [1122, 218], [1122, 358], [891, 3...   
1   [[102, 300], [818, 300], [818, 895], [102, 895]]   

                                                   1  
0                               Apollo LPHARMACY gsk  
1  R AMOXYCILLIN AND POTASSIUM CLAVULANATE TABLET...

In [7]:
text3 = text[1][1]

In [8]:
text3

'R AMOXYCILLIN AND POTASSIUM CLAVULANATE TABLETS IP AUGMENTIN 375 1 Co-Tabetclav Each film-coated tablet contains: Amoxycillin Trihydrate IP equivalent to Amoxycillin 250 mg Potassium Clavulanate Diluted IP equivalent to Clavulanic Acid 125 mg Colour: Titanium Dioxide IP To be used as directed by physician Tablet to be consumed in whole, not t0 be broken 3 Store protected from moisture at a temperature not exceeding 25*C . Keep out of reach of children For Toll free Customer Care: Call 1800222203 Original Research Product of SmithKline Beecham Plc, London:'

In [9]:
!pip install simpletransformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 250.5/250.5 KB 8.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 KB 5.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 462.8/462.8 KB 32.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 77.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 76.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 90.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.6/9.6 MB 97.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 98.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 190.3/190.3 KB 23.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 184.3/184.3 KB 24.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━

In [10]:
import spacy
from wordcloud import WordCloud, STOPWORDS
from spacy.util import minibatch, compounding

In [11]:
import matplotlib.pyplot as plt
import re
import random

LOADING DATASET


In [13]:
from google.colab import drive
drive.mount("/content/gdrive/")

Mounted at /content/gdrive/


In [14]:
import pandas as pd
df = pd.read_csv("/content/gdrive/MyDrive/healthh/train_Dataset_health.csv")

In [15]:
df

uniqueID                  drugName                     condition  \
0         206461                 Valsartan  Left Ventricular Dysfunction   
1          95260                Guanfacine                          ADHD   
2          92703                    Lybrel                 Birth Control   
3         138000                Ortho Evra                 Birth Control   
4          35696  Buprenorphine / naloxone             Opiate Dependence   
...          ...                       ...                           ...   
161292    191035                   Campral            Alcohol Dependence   
161293    127085            Metoclopramide               Nausea/Vomiting   
161294    187382                   Orencia          Rheumatoid Arthritis   
161295     47128        Thyroid desiccated           Underactive Thyroid   
161296    215220              Lubiprostone         Constipation, Chronic   

                                                   review  rating       date  \
0       "It has no side effect, I take it in combinati...       9  20-May-12   
1       "My son is halfway through his fourth week of ...       8  27-Apr-10   
2       "I used to take another oral contraceptive, wh...       5  14-Dec-09   
3       "This is my first time using any form of birth...       8  03-Nov-15   
4       "Suboxone has completely turned my life around...       9  27-Nov-16   
...                                                   ...     ...        ...   
161292  "I wrote my first report in Mid-October of 201...      10  31-May-15   
161293  "I was given this in IV before surgey. I immed...       1  01-Nov-11   
161294  "Limited improvement after 4 months, developed...       2  15-Mar-14   
161295  "I&#039;ve been on thyroid medication 49 years...      10  19-Sep-15   
161296  "I&#039;ve had chronic constipation all my adu...       9  13-Dec-14   

        usefulCount  
0                27  
1               192  
2                17  
3                10  
4                37  
...             ...  
161292          125  
161293           34  
161294           35  
161295           79  
161296          116  

[161297 rows x 7 columns]

PREPROCESSING


In [16]:
# Loading pre-existing spacy model
import spacy
nlp=spacy.load('en_core_web_sm')

# Initializing pipeline
ner=nlp.get_pipe("ner")

In [17]:
#Tokenization
def process_review(review):
    processed_token = []
    for token in review.split():
        token = ''.join(e.lower() for e in token if e.isalnum())
        processed_token.append(token)
    return ' '.join(processed_token)

In [18]:
#Drugs Names
all_drugs = df['drugName'].unique().tolist()

In [19]:
all_drugs = [x.lower() for x in all_drugs]

In [20]:
all_drugs

['valsartan',
 'guanfacine',
 'lybrel',
 'ortho evra',
 'buprenorphine / naloxone',
 'cialis',
 'levonorgestrel',
 'aripiprazole',
 'keppra',
 'ethinyl estradiol / levonorgestrel',
 'topiramate',
 'l-methylfolate',
 'pentasa',
 'dextromethorphan',
 'nexplanon',
 'liraglutide',
 'trimethoprim',
 'amitriptyline',
 'lamotrigine',
 'nilotinib',
 'atripla',
 'trazodone',
 'etonogestrel',
 'etanercept',
 'tioconazole',
 'azithromycin',
 'eflornithine',
 'daytrana',
 'ativan',
 'imitrex',
 'sertraline',
 'toradol',
 'viberzi',
 'mobic',
 'dulcolax',
 'morphine',
 'moviprep',
 'trilafon',
 'fluconazole',
 'contrave',
 'clonazepam',
 'metaxalone',
 'venlafaxine',
 'ledipasvir / sofosbuvir',
 'symbyax',
 'tamsulosin',
 'doxycycline',
 'dulaglutide',
 'intuniv',
 'buprenorphine',
 'qvar',
 'opdivo',
 'pyridium',
 'latuda',
 'bupropion',
 'implanon',
 'effexor xr',
 'drospirenone / ethinyl estradiol',
 'nuvaring',
 'prepopik',
 'tretinoin',
 'gildess fe 1 / 20',
 'ethinyl estradiol / norgestimate'

In [21]:
count = 0
TRAIN_DATA = []
for _, item in df.iterrows():
    ent_dict = {}
    if count < 1000:
        review = process_review(item['review'])
        visited_items = []
        entities = []
        for token in review.split():
            if token in all_drugs:
                for i in re.finditer(token, review):
                    if token not in visited_items:
                        entity = (i.span()[0], i.span()[1], 'DRUG')
                        visited_items.append(token)
                        entities.append(entity)
        if len(entities) > 0:
            ent_dict['entities'] = entities
            train_item = (review, ent_dict)
            TRAIN_DATA.append(train_item)
            count+=1

In [32]:
TRAIN_DATA

[('i started using levora after stopping yasmin i039m starting my third levora pack next week and so far my experience with levora has been great no headaches no nausea no feeling physically weak or overly emotional my libido is really high and my breasts have gotten a little bigger i did however have the food cravings as mentioned in other comments and it did make me gain a little weight but i notice that when i039m taking the reminder pills i039m a lot slimmer than when i039m taking the active pills i also have increased acne by very very little nothing that makes me complain or feel annoyed i recommend levora for anyone debating on it it worked well for me',
  {'entities': [(16, 22, 'DRUG'), (38, 44, 'DRUG')]}),
 ('i wish this was around back when i was in high school i have struggled with acne for my whole teenadult life i have taken yaz on and off for 8 years now and my skin never looks better than when i039m on yaz the doc took me off last year because i was getting headaches and

**LABELLING**

In [22]:
# Adding label 'DRUG' to the ner model
for _, annotations in TRAIN_DATA:
  for ent in annotations.get("entities"):
    ner.add_label(ent[2])

**TRAINING**

In [23]:
# Disable pipeline components you dont need to change
pipe_exceptions = ["ner", "trf_wordpiecer", "trf_tok2vec"]
unaffected_pipes = [pipe for pipe in nlp.pipe_names if pipe not in pipe_exceptions]

In [24]:
# Import requirements
from spacy.training.example import Example
import random
from spacy.util import minibatch, compounding
from pathlib import Path
def train_ner(TRAIN_DATA):
  # TRAINING THE MODEL
  with nlp.disable_pipes(*unaffected_pipes):

    # Training for 15 iterations
    for iteration in range(15):

      # shuffles examples before every iteration for better accuracy
      random.shuffle(TRAIN_DATA)
      losses = {}
      # batch up the examples using spaCy's minibatch(used for dividing the training_set into mini batches for training)
      batches = minibatch(TRAIN_DATA, size=compounding(4.0, 32.0, 1.001))
      for batch in batches:
        texts, annotations = zip(*batch)
        example = []
        # Update the model with iterating each text (Optimizer)
        for i in range(len(texts)):
          doc = nlp.make_doc(texts[i])
          example.append(Example.from_dict(doc, annotations[i]))
      # Update the model (to learn from losses)
        nlp.update(example, drop=0.5, losses=losses)
        print("Losses",losses)
  return nlp

In [25]:
nlp2 = train_ner(TRAIN_DATA) 

Losses {'ner': 34.34911864933785}
Losses {'ner': 52.117288484657706}
Losses {'ner': 75.37337072341015}
Losses {'ner': 89.45035483228187}
Losses {'ner': 104.45828370781808}
Losses {'ner': 124.44103543597976}
Losses {'ner': 144.43551483247845}
Losses {'ner': 155.02535171764123}
Losses {'ner': 166.32706291332133}
Losses {'ner': 179.91869204851795}
Losses {'ner': 191.96169296420433}
Losses {'ner': 208.26908471907453}
Losses {'ner': 222.02767683298217}
Losses {'ner': 234.4198226751569}
Losses {'ner': 244.41123038025742}
Losses {'ner': 258.6857771577686}
Losses {'ner': 269.04903770936505}
Losses {'ner': 278.92708151127465}


/usr/local/lib/python3.8/dist-packages/spacy/training/iob_utils.py:149: UserWarning: [W030] Some entities could not be aligned in the text "i am on my third day of taking 50mg of pristiq aft..." with entities "[(39, 46, 'DRUG'), (81, 88, 'DRUG')]". Use `spacy.training.offsets_to_biluo_tags(nlp.make_doc(text), entities)` to check the alignment. Misaligned entities ('-') will be ignored during training.
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/spacy/training/iob_utils.py:149: UserWarning: [W030] Some entities could not be aligned in the text "i received this from the pharmacy a couple of diff..." with entities "[(108, 114, 'DRUG'), (223, 229, 'DRUG'), (701, 707...". Use `spacy.training.offsets_to_biluo_tags(nlp.make_doc(text), entities)` to check the alignment. Misaligned entities ('-') will be ignored during training.
  warnings.warn(


Losses {'ner': 294.0151929551911}
Losses {'ner': 306.52144606690115}
Losses {'ner': 316.86736982098216}
Losses {'ner': 328.58929755740417}
Losses {'ner': 336.5749287108573}
Losses {'ner': 344.91834183113485}
Losses {'ner': 365.93538427621195}
Losses {'ner': 375.7123355458515}
Losses {'ner': 387.52757540387927}
Losses {'ner': 399.2077245688323}
Losses {'ner': 409.09024934461}
Losses {'ner': 416.6291324829416}
Losses {'ner': 427.10952607698516}
Losses {'ner': 440.4598584774757}
Losses {'ner': 454.0936664184745}
Losses {'ner': 461.79142113370915}
Losses {'ner': 475.19789921830386}
Losses {'ner': 482.94112712574577}
Losses {'ner': 490.6354990288281}
Losses {'ner': 500.37632324778394}
Losses {'ner': 509.7317584318142}
Losses {'ner': 523.1192617746282}
Losses {'ner': 540.0626151653422}
Losses {'ner': 552.8276310301434}
Losses {'ner': 565.1506668727333}
Losses {'ner': 577.9820954988753}
Losses {'ner': 587.9253875042922}
Losses {'ner': 595.6944818906376}
Losses {'ner': 606.0168055609428}
Losse

/usr/local/lib/python3.8/dist-packages/spacy/training/iob_utils.py:149: UserWarning: [W030] Some entities could not be aligned in the text "had severe diarrhea admitted in hospital they gave..." with entities "[(54, 60, 'DRUG')]". Use `spacy.training.offsets_to_biluo_tags(nlp.make_doc(text), entities)` to check the alignment. Misaligned entities ('-') will be ignored during training.
  warnings.warn(


Losses {'ner': 1398.1103536999478}
Losses {'ner': 1416.1988591203667}
Losses {'ner': 1431.0306382751592}
Losses {'ner': 1435.603056320077}
Losses {'ner': 1442.9695483971925}
Losses {'ner': 1448.713137689368}
Losses {'ner': 1457.0306611367814}
Losses {'ner': 1465.3081961899566}
Losses {'ner': 1474.2147135541338}
Losses {'ner': 1482.7010394328447}
Losses {'ner': 1494.9140511061746}
Losses {'ner': 1503.6754794133294}
Losses {'ner': 1507.2429897514357}
Losses {'ner': 1510.134708645669}
Losses {'ner': 1520.469015236107}
Losses {'ner': 1526.8520782490714}
Losses {'ner': 1535.5731555355603}
Losses {'ner': 1543.1330474061397}
Losses {'ner': 1548.5752445403925}
Losses {'ner': 1553.8286610178225}
Losses {'ner': 1561.2237205639028}
Losses {'ner': 1566.6732882048868}
Losses {'ner': 1573.7742102842942}
Losses {'ner': 1584.7756797011498}
Losses {'ner': 1591.7028758643678}
Losses {'ner': 1597.5324112042283}


/usr/local/lib/python3.8/dist-packages/spacy/training/iob_utils.py:149: UserWarning: [W030] Some entities could not be aligned in the text "awful medicine the worst the side effects outweigh..." with entities "[(178, 183, 'DRUG'), (342, 350, 'DRUG')]". Use `spacy.training.offsets_to_biluo_tags(nlp.make_doc(text), entities)` to check the alignment. Misaligned entities ('-') will be ignored during training.
  warnings.warn(


Losses {'ner': 1601.7877604319624}
Losses {'ner': 1607.337236968333}
Losses {'ner': 1613.6450099810565}
Losses {'ner': 1619.0694143756932}
Losses {'ner': 1630.8254716804754}
Losses {'ner': 1647.2243058347565}
Losses {'ner': 1654.8950466289953}
Losses {'ner': 1662.6779089067106}
Losses {'ner': 1670.5876995183226}
Losses {'ner': 1678.266135896629}
Losses {'ner': 1685.3298347944947}
Losses {'ner': 1694.5762713783786}
Losses {'ner': 1700.4281264405488}
Losses {'ner': 1704.4544837358126}
Losses {'ner': 1710.104943577261}
Losses {'ner': 1713.4294802304535}
Losses {'ner': 1722.3968554878727}
Losses {'ner': 1727.6615448733685}
Losses {'ner': 1735.3257377100604}
Losses {'ner': 1744.5493745763442}
Losses {'ner': 1749.137303896797}
Losses {'ner': 1757.3700490137503}
Losses {'ner': 1763.0680594476426}
Losses {'ner': 1771.3893693517025}
Losses {'ner': 1779.9827855436988}
Losses {'ner': 1783.4710236502472}
Losses {'ner': 1790.398958240105}
Losses {'ner': 1793.1237753690139}
Losses {'ner': 1801.86704

/usr/local/lib/python3.8/dist-packages/spacy/training/iob_utils.py:149: UserWarning: [W030] Some entities could not be aligned in the text "many people are unreasonably scared of klonopin039..." with entities "[(39, 47, 'DRUG')]". Use `spacy.training.offsets_to_biluo_tags(nlp.make_doc(text), entities)` to check the alignment. Misaligned entities ('-') will be ignored during training.
  warnings.warn(


Losses {'ner': 2027.7092178315536}
Losses {'ner': 2036.338463779621}
Losses {'ner': 2043.4750417461719}
Losses {'ner': 2052.945010537845}
Losses {'ner': 2067.218313648707}
Losses {'ner': 2070.3440036750967}
Losses {'ner': 2076.7248961805603}
Losses {'ner': 2080.4101760278436}
Losses {'ner': 2085.375523441981}
Losses {'ner': 2091.2818122526583}
Losses {'ner': 2099.080107634726}
Losses {'ner': 2106.4001458505186}
Losses {'ner': 2116.315431254024}
Losses {'ner': 2125.640746356187}
Losses {'ner': 2135.6415966001528}
Losses {'ner': 2140.482481357466}
Losses {'ner': 2147.504841716537}
Losses {'ner': 2157.3488506343783}
Losses {'ner': 2165.169086065524}
Losses {'ner': 2172.648253293435}
Losses {'ner': 2181.2699007240776}
Losses {'ner': 2195.9951943849646}
Losses {'ner': 2205.3517075378145}
Losses {'ner': 2212.608643652177}
Losses {'ner': 2221.5473280836773}
Losses {'ner': 2234.073738194149}
Losses {'ner': 2246.484050343456}
Losses {'ner': 2258.042434178421}
Losses {'ner': 2263.6651978808604}


**METRICS**

In [26]:
import spacy
from spacy.util import minibatch
from spacy.scorer import Scorer
from tqdm import tqdm
import random
def accuracy_score(test_data):    
        examples = []
        scorer = Scorer()
        for text, annotations in test_data:
            pred_doc = nlp2(text)
            try:
                example = Example.from_dict(pred_doc, annotations)
            except:
                print(f'Error: failed to process document: \n{text},\n\n annotations: {annotations}')
                continue
            examples.append(example)
            
        accuracy = scorer.score(examples)
        
        return accuracy       
        
    

In [27]:
accuracy_score(TRAIN_DATA[:200])

{'token_acc': 1.0,
 'token_p': 1.0,
 'token_r': 1.0,
 'token_f': 1.0,
 'sents_p': None,
 'sents_r': None,
 'sents_f': None,
 'tag_acc': None,
 'pos_acc': None,
 'morph_acc': None,
 'morph_micro_p': None,
 'morph_micro_r': None,
 'morph_micro_f': None,
 'morph_per_feat': None,
 'dep_uas': None,
 'dep_las': None,
 'dep_las_per_type': None,
 'ents_p': 0.9324324324324325,
 'ents_r': 0.968421052631579,
 'ents_f': 0.9500860585197936,
 'ents_per_type': {'DRUG': {'p': 0.9324324324324325,
   'r': 0.968421052631579,
   'f': 0.9500860585197936}},
 'cats_score': 0.0,
 'cats_score_desc': 'macro F',
 'cats_micro_p': 0.0,
 'cats_micro_r': 0.0,
 'cats_micro_f': 0.0,
 'cats_macro_p': 0.0,
 'cats_macro_r': 0.0,
 'cats_macro_f': 0.0,
 'cats_macro_auc': 0.0,
 'cats_f_per_type': {},
 'cats_auc_per_type': {}}

**RESULT**

{'DRUG': {'p': 0.9324324324324325,
   'r': 0.968421052631579,
   'f': 0.9500860585197936}}

   f1_score (accuracy): 0.9500860585197936  (95.0%)
   recall: 0.968421052631579
   precision: 0.9324324324324325



SAVING MODEL


In [28]:
model = None
output_dir=Path('/content/model')
if output_dir is not None:
    output_dir = Path(output_dir)
    if not output_dir.exists():
        output_dir.mkdir()
    nlp.to_disk(output_dir)
    print("Saved model to", output_dir)

Saved model to /content/model


Displaying the name of medicine along with it's description

In [29]:
doc = nlp2(text3)
for ent in doc.ents:
  print(ent,end = " ")
print('\n')
print(text3,'\n')

Amoxycillin 

R AMOXYCILLIN AND POTASSIUM CLAVULANATE TABLETS IP AUGMENTIN 375 1 Co-Tabetclav Each film-coated tablet contains: Amoxycillin Trihydrate IP equivalent to Amoxycillin 250 mg Potassium Clavulanate Diluted IP equivalent to Clavulanic Acid 125 mg Colour: Titanium Dioxide IP To be used as directed by physician Tablet to be consumed in whole, not t0 be broken 3 Store protected from moisture at a temperature not exceeding 25*C . Keep out of reach of children For Toll free Customer Care: Call 1800222203 Original Research Product of SmithKline Beecham Plc, London: 



TESTING ON CHUNKS OF DATA FROM TRAINING **SET** 

In [30]:
#Testing model on our training data
for text,_ in TRAIN_DATA[:5]:
  doc = nlp2(text)
  result = [(ent) for ent in doc.ents]
  print(result)

[levora]
[yaz]
[peppermint]
[topamax, sumatriptan]
[caffeine]


In [31]:
#Loading training result from model folder and testing on training set
print("Loading from", output_dir)
nlp2 = spacy.load(output_dir)
for text, _ in TRAIN_DATA[10:25]:
    doc = nlp2(text)
    print('Entities', [(ent.text, ent.label_) for ent in doc.ents])
    print('Tokens', [(t.text, t.ent_type_, t.ent_iob) for t in doc])

Loading from /content/model
Entities [('ibuprofen', 'DRUG')]
Tokens [('honestly', '', 2), ('i', '', 2), ('felt', '', 2), ('better', '', 2), ('about', '', 2), ('4', '', 2), ('hours', '', 2), ('after', '', 2), ('begging', '', 2), ('treatment', '', 2), ('with', '', 2), ('this', '', 2), ('for', '', 2), ('an', '', 2), ('infectedabcessed', '', 2), ('tooth', '', 2), ('surprisingly', '', 2), ('quick', '', 2), ('considering', '', 2), ('the', '', 2), ('amount', '', 2), ('of', '', 2), ('pain', '', 2), ('i', '', 2), ('was', '', 2), ('in', '', 2), ('the', '', 2), ('downside', '', 2), ('for', '', 2), ('me', '', 2), ('at', '', 2), ('least', '', 2), ('seems', '', 2), ('to', '', 2), ('be', '', 2), ('slight', '', 2), ('nauseadizziness', '', 2), ('i', '', 2), ('had', '', 2), ('taken', '', 2), ('ibuprofen', 'DRUG', 3), ('earlier', '', 2), ('due', '', 2), ('to', '', 2), ('pain', '', 2), ('so', '', 2), ('this', '', 2), ('meant', '', 2), ('bad', '', 2), ('times', '', 2), ('in', '', 2), ('front', '', 2), ('of